File: packages/jupyter-scales/src/continuous.ts
// Copyright (c) Jupyter Development Team.
// Distributed under the terms of the Modified BSD License.

import {
  WidgetModel, ManagerBase
} from '@jupyter-widgets/base';

import {
  InterpolatorFactory, scaleLinear, scaleLog, scalePow, ScaleLinear
} from 'd3-scale';

import * as d3Interpolate from 'd3-interpolate';

import {
  JUPYTER_EXTENSION_VERSION
} from './version';

import {
  ScaleModel
} from './scale';


/**
 * Find the name of the d3-interpolate function
 */
function interpolatorName(fun: Function) {
  for (let key of Object.keys(d3Interpolate)) {
    if ((d3Interpolate as any)[key] === fun) {
      return key;
    }
  }
  throw ReferenceError(`Cannot find name of interpolator ${fun}`);
}

/**
 * A widget model of a continuous scale
 */
export abstract class ContinuousScaleModel extends ScaleModel {
  defaults() {
    return {...super.defaults(),
      domain: [0, 1],
      range: [0, 1],

      interpolator: 'interpolate',
      clamp: false,
    };
  }

  createPropertiesArrays() {
    super.createPropertiesArrays();
    this.simpleProperties.push(
      'domain',
      'range',
      'clamp',
    );
  }

  /**
   * Create the wrapped d3-scale scaleLinear object
   */
  abstract constructObject(): any;

  /**
   * Sync the model properties to the d3 object.
   */
  syncToObject() {
    super.syncToObject();
    let interpolatorName = this.get('interpolator') || 'interpolate';
    let interpolator = (d3Interpolate as any)[interpolatorName] as InterpolatorFactory<number, number>;
    this.obj.interpolate(interpolator);
  }

  /**
   * Synt the d3 object properties to the model.
   */
  syncToModel(toSet: Backbone.ObjectHash) {
    let interpolator = this.obj.interpolate() as InterpolatorFactory<any, any>;
    toSet['interpolator'] = interpolatorName(interpolator);
    super.syncToModel(toSet);
  }

  static serializers = {
    ...ScaleModel.serializers,
  }
}


/**
 * A widget model of a linear scale
 */
export
class LinearScaleModel extends ContinuousScaleModel {

  /**
   * Create the wrapped d3-scale scaleLinear object
   */
  constructObject(): any {
    return scaleLinear();
  }

  obj: ScaleLinear<any, any>;

  static serializers = {
    ...ContinuousScaleModel.serializers,
  }

  static model_name = 'LinearScaleModel';
}


/**
 * A widget model of a linear scale
 */
export
class LogScaleModel extends ContinuousScaleModel {
  defaults() {
    return {...super.defaults(),
      base: 10,
    };
  }

  createPropertiesArrays() {
    super.createPropertiesArrays();
    this.simpleProperties.push(
      'base',
    );
  }

  /**
   * Create the wrapped d3-scale scaleLinear object
   */
  constructObject(): any {
    return scaleLog();
  }

  static serializers = {
    ...ContinuousScaleModel.serializers,
  }

  static model_name = 'LogScaleModel';
}


/**
 * A widget model of a linear scale
 */
export
class PowScaleModel extends ContinuousScaleModel {
  defaults() {
    this.constructor
    return {...super.defaults(),
      exponent: 1,
    };
  }

  createPropertiesArrays() {
    super.createPropertiesArrays();
    this.simpleProperties.push(
      'exponent',
    );
  }

  /**
   * Create the wrapped d3-scale scaleLinear object
   */
  constructObject(): any {
    return scalePow();
  }

  static serializers = {
    ...ContinuousScaleModel.serializers,
  }

  static model_name = 'PowScaleModel';
}

File: ipyscales/continuous.py
#!/usr/bin/env python
# coding: utf-8

# Copyright (c) Jupyter Development Team.
# Distributed under the terms of the Modified BSD License.

"""
Defines linear scale widget, and any supporting functions
"""

from traitlets import Float, CFloat, Unicode, List, Union, Bool, Any
from ipywidgets import Color, register

from .scale import Scale
from .traittypes import VarlenTuple


class ContinuousScale(Scale):
    """A continuous scale widget.

    This should be treated as an abstract class, and should
    not be directly instantiated.
    """

    domain = VarlenTuple(trait=CFloat(), default_value=(0.0, 1.0), minlen=2).tag(
        sync=True
    )

    range = VarlenTuple(trait=Any(), default_value=(0.0, 1.0), minlen=2).tag(sync=True)

    interpolator = Unicode("interpolate").tag(sync=True)
    clamp = Bool(False).tag(sync=True)


@register
class LinearScale(ContinuousScale):
    """A linear scale widget.

    See the documentation for d3-scale's linear for
    further details.
    """

    _model_name = Unicode("LinearScaleModel").tag(sync=True)


@register
class LogScale(ContinuousScale):
    """A logarithmic scale widget.

    See the documentation for d3-scale's scaleLog for
    further details.
    """

    _model_name = Unicode("LogScaleModel").tag(sync=True)

    domain = VarlenTuple(trait=CFloat(), default_value=(1.0, 10.0), minlen=2).tag(
        sync=True
    )

    base = Float(10).tag(sync=True)


@register
class PowScale(ContinuousScale):
    """A power scale widget.

    See the documentation for d3-scale's scaleLog for
    further details.
    """

    _model_name = Unicode("PowScaleModel").tag(sync=True)

    exponent = Float(1).tag(sync=True)
